### PR TITLE
RFB 2: UI Structure for Heterogeneous Signer Support

### PR DESCRIPTION
**Pull request by Shadowfiend**
Quoting from the proposal section:

> This RFB focuses on a scalable approach to architecting the UI flow for
signing, irrespective of whether it is transaction or message data being
signed. It does not speak to the interaction between transaction and message
data, nor to the handling of multiple pending signing requests (something that
has been informally referred to as "transaction queueing"), which is treated as
an orthogonal concern[.]

### CODE REVIEW

**Comment by mhluongo** [comment]
Clarifying the RFB process here... do you want discussions in this PR, or a light review & merge, then discuss?

I lean toward in-PR discussions but have done it both ways.

**Comment by Shadowfiend** [comment]
Discussion in the PR for sure.

**Comment by mhluongo** [comment]
I think your feedback is especially important here @VladUXUI — this is as close to an accurate mental model as I've seen for the different types of signing across other wallets and chains, and aligning the UX with that accurate model is going to be a big unlock for user education and signing safety.

File: rfb/rfb-2-signers-ui.md [hunk 1]
```diff
@@ -0,0 +1,610 @@
...
+      SignerBaseFrame.tsx
+      SignerReadOnly/
+        SignerReadOnlyFrame.tsx
+        SignerReadOnlySigning.tsx
+      SignerKeyring/
+        SignerKeyringFrame.tsx
+        SignerKeyringSigning.tsx
+      SignerLedger/
+        SignerLedgerFrame.tsx
+        SignerLedgerSigning.tsx
```
Comment (henryboldi) [inline rfb/rfb-2-signers-ui.md:180]:
```suggestion
components/
  Sign/
    index.tsx
    SignData/
      SignData.tsx
      SignDataTransaction/
        ...
      SignDataMessage/
        ...
    Signer/
      SignerBaseFrame.tsx
      SignerReadOnly/
        SignerReadOnlyFrame.tsx
        SignerReadOnlySigning.tsx
      SignerKeyring/
        SignerKeyringFrame.tsx
        SignerKeyringSigning.tsx
      SignerLedger/
        SignerLedgerFrame.tsx
        SignerLedgerSigning.tsx
```

The meaning of `Signing` here might be getting muddled:
- `SigningDataMessage` is a UI components that display analysis
- `Signing/` contains all components sign related
- `SignerLedgerSigning` is a component that chiefly dispatches the sign action in an essentially declarative way

How about this suggestion?

File: rfb/rfb-2-signers-ui.md [hunk 2]
```diff
@@ -0,0 +1,610 @@
...
+/**
+ * The React component type of a signing frame; all *Frame components in
+ * subdirectories should conform to this signature, enforced by the
+ * frameComponentForSigner lookup.
+ */
+export type SigningFrame = (props: SigningFrameProps) => ReactElement
+
+// Takes a signing request and resolves the signer that should be used to sign
+// it and the details of signing data for user presentation.
+function resolveSignatureDetails(request: SigningRequest): ResolvedSignatureDetails {
```
Comment (henryboldi) [inline rfb/rfb-2-signers-ui.md:262]: Why call this in the component rather than having it resolved ahead of time?

File: rfb/rfb-2-signers-ui.md [hunk 3]
```diff
@@ -0,0 +1,610 @@
...
+- The details resolver (`resolveSignatureDetails` and its delegates) is similar to the
+  `SignTransactionInfoProvider`, but is deliberately oriented towards resolving data,
+  and thus not structured as a component but as a plain function. Instead,
+  component-related decisions are left to the signer frame.
+- The `Signing` container behaves similarly to the `SignTransactionContainer`,
+  but, as with the details resolver, leaves things like the review panel to the
+  signer frame.
+- Much of the conditional logic that handles Ledger is left entirely to the
+  Ledger frame instead of being entwined with the shared transaction analysis
+  logic. As a specific example, the `reviewPanel` used in the current
```
Comment (henryboldi) [inline rfb/rfb-2-signers-ui.md:552]: Beautiful!

File: rfb/rfb-2-signers-ui.md [hunk 4]
```diff
@@ -0,0 +1,610 @@
...
+  expect_, and is instead focused on helping the user confirm that the
+  transaction in the wallet is the same one they are signing on their hardware
+  device.
+
+Keychain unlocking in particular diverges somewhat from the current functioning:
+
+- The flow of keychain unlock moves from the beginning of the signing flow
+  (before transaction data is presented) to the end of it (once the user
+  chooses to sign). This aligns keychain signing with what is expected to
+  happen in other signers, which generally don't block on signer availability
```
Comment (henryboldi) [inline rfb/rfb-2-signers-ui.md:566]: https://github.com/tallycash/extension/issues/1389 👀

File: rfb/rfb-2-signers-ui.md [hunk 5]
```diff
@@ -0,0 +1,610 @@
...
+is not always present, and other tools like Sign In with Ethereum produce
+signatures that aren't necessarily used for on-chain actions.
+
+This RFB focuses specifically on the question of how to structure UI component
+flow to accommodate different types of signers, but _does not_ look at how
+different signature types should flow between each other.
+
+## Related Links
+
+- [EIP-191: Signed Data Standard](eip191)
```
Comment (jagodarybacka) [inline rfb/rfb-2-signers-ui.md:604]: These links are not working for me 😞

**Comment by jagodarybacka** [comment]
Looks very good 🥇  
I can't wait to see it implemented, even by looking at the folders and files structure it is much simpler to understand what is going on with transactions.

**Comment by SJQuality** [comment]
QA will need some explanation on how to test this. I'm not sure what this is. :)

**Comment by mhluongo** [comment]
It's an [RFC](https://www.techtarget.com/whatis/definition/Request-for-Comments-RFC)-inspires architectural doc @SJQuality! No need to test the Markdown, unless you want to drop some feedback :)

File: rfb/rfb-2-signers-ui.md (hunk 4 above)
Comment (mhluongo) [reply rfb/rfb-2-signers-ui.md:566]: Yep we've already been stung

File: rfb/rfb-2-signers-ui.md [hunk 6]
```diff
@@ -0,0 +1,610 @@
...
+    if (!signingInitiated && keyringStatus === "unlocked") {
+      dispatch(signActionCreator())
+
+      setSigningInitiated(true)
+    }
+  }, [keyringStatus, signingInitiated, setSigningInitiated])
+
+  // In this construction, keyring unlocking isn't done as a route, but in line
+  // in the signing frame.
+  if (keyringStatus === "uninitialized") {
```
Comment (greg-nagy) [inline rfb/rfb-2-signers-ui.md:427]: O.O 
neat

File: rfb/rfb-2-signers-ui.md [hunk 7]
```diff
@@ -0,0 +1,610 @@
...
+        ...
+      SigningDataMessage/
+        ...
+    Signer/
+      SignerBaseFrame.tsx
+      SignerReadOnly/
+        SignerReadOnlyFrame.tsx
+        SignerReadOnlySigning.tsx
+      SignerKeyring/
+        SignerKeyringFrame.tsx
```
Comment (greg-nagy) [inline rfb/rfb-2-signers-ui.md:176]: Folder and file nameing is a bit redundant.  I would choose either to have a flat folder structure and keep the file nameing or remove the folder name from the beginning of the filename.

**Review by greg-nagy — APPROVED** [review]
I reallz love this RFB, and would like to have more :) 
Will need to speed up their review though.

File: rfb/rfb-2-signers-ui.md [hunk 8]
```diff
@@ -0,0 +1,610 @@
...
+    index.tsx
+    SigningData/
+      SigningData.tsx
+      SigningDataTransaction/
+        ...
+      SigningDataMessage/
+        ...
+    Signer/
+      SignerBaseFrame.tsx
+      SignerReadOnly/
```
[the remark this answers is not in the document]
Comment (mhluongo) [reply rfb/rfb-2-signers-ui.md:172]: It does though! Read-only accounts ideally still have a signing interface, they just don't use key material to sign... if the user can come up with a signature, so much the better.

This lets us handle airgapped setups and whatnot.

File: rfb/rfb-2-signers-ui.md (hunk 7 above)
Comment (Shadowfiend) [reply rfb/rfb-2-signers-ui.md:176]: As far as I can tell, doing naming this way seems to be a standard react pattern, and I don't want to mess with it. It's not my preferred approach either 😬

File: rfb/rfb-2-signers-ui.md [hunk 9]
```diff
@@ -0,0 +1,610 @@
...
+||                 ||
+|| --------------- ||
+||  signer footer  ||
+|| --------------- ||
+---------------------
+```
+
+The `Signing` component is charged with resolving the appropriate
+`SigningData`, which renders the UI for analysis of transaction or message
+data, as well as the `SignerFrame`, which is the frame component specific to
```
[the remark this answers is not in the document]
Comment (Shadowfiend) [reply rfb/rfb-2-signers-ui.md:140]: I don't remember exactly how I landed on it, but I would say the closest analog that is named the same way are probably HTML frames.

File: rfb/rfb-2-signers-ui.md [hunk 10]
```diff
@@ -0,0 +1,610 @@
...
+### Goal
+
+Once this refactor is complete, the intent is to have a clear separation
+between data understanding concerns and signer-specific concerns, and to
+provide infrastructure for contributors to add UI support for new signer types
+as needed. The signer-specific areas of the code will need to control the frame
+of the signing action (the UI around the signature data) while the user is in
+the analysis phase and, once the user has initiated signing, the full UI.
+
+New signer types may need their own redux slices, their own component flows
```
[the remark this answers is not in the document]
Comment (Shadowfiend) [reply rfb/rfb-2-signers-ui.md:111]: I lived this journey when I wrote the light RFC for signing heh. There may still be a factoring that reduced the need for an entire new slice for signers, but it's going to require some work. If it becomes apparent, definitely worth its own RFB.

File: rfb/rfb-2-signers-ui.md [hunk 11]
```diff
@@ -0,0 +1,610 @@
...
+Keystone, to remote wallet protocols like Wallet Connect, the wallet is
+eventually meant to allow arbitrary signers to sign data that can produce
+outcomes on various cryptocurrency chains.
+
+To this end, the architecture of the UI flow around signing needs to be adapted
+to the variations between different signers, while still supporting a common
+core of functionality across signing tasks, such as presenting enriched
+versions of the data being signed (for user clarity and security) and allowing
+access to underlying data and, for transactions, transaction details such as
+network fees.
```
[the remark this answers is not in the document]
Comment (Shadowfiend) [reply rfb/rfb-2-signers-ui.md:18]: Does this feel like something that could be handled in the signer frame, or would it need to be handled elsewhere?

File: rfb/rfb-2-signers-ui.md (hunk 5 above)
Comment (Shadowfiend) [reply rfb/rfb-2-signers-ui.md:604]: Yep, got this format slightly wrong 👀

File: rfb/rfb-2-signers-ui.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (greg-nagy) [reply rfb/rfb-2-signers-ui.md:18]: Signer frame is one part of the solution the other one is ledger plugins so we can decode transactions on the ledger, provider more info on it's screen (and hopefully remove the need for enabling blind signing)

If/When we will have the ledger plugin the Signer frame need to mimic/follow the changed behavior.

File: rfb/rfb-2-signers-ui.md [hunk 12]
```diff
@@ -62,10 +62,10 @@ A few terms are used below and are worth defining:
   outside the memory space of the extension; may be remote via USB, the
   internet, or even air gap/QR code.
 - **Personal message signing**: Signing of unstructured plain text messages,
-  specified by [EIP-191](eip191). Unstructured messages are prefixed by a
+  specified by [EIP-191][eip191]. Unstructured messages are prefixed by a
```
Comment (greg-nagy) [inline rfb/rfb-2-signers-ui.md:65]: Didn't know we can do this, but this is cool!

**Review by greg-nagy — APPROVED** [review]
Let's merge this 💪

If we need to update it in the future we can submit a new PR

File: rfb/rfb-2-signers-ui.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (Shadowfiend) [reply rfb/rfb-2-signers-ui.md:18]: Yeah, definitely agree on that front.

**Pull request merged**